[daitenshionyan] iP

daitenshionyan opened the pull request:
### Aqua
> "Your mind is for having ideas, not holding them." – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

Aqua frees your mind of having to remember things you need to do. She is,
* Text-based
* Easy to learn
* ~~FAST~~ _SUPER_ FAST to use

All you need to do is,
1. Download it from [here](https://github.com/daitenshionyan/ip/releases).
2. Double-click it.
3. Add your tasks.
4. Let it manage your tasks for you 😉

And she is **CUTE**!

Features:
- [x] Managing tasks
- [ ]  Managing deadlines (coming soon)
- [ ]  Reminders (coming soon)
________________________________________
If you are a Java programmer, you can use it to practice Java too. Here is the `main` method:
```java
public class Aqua {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Aqua: fix follow up executing on failure

Commit:
EXPECTED: correct

Commit:
Understand dates and times

Commit:
Add custom date time parser

Commit:
AquaTask: format printed dates

Commit:
DateUtils: fix time parsing

Commit:
AquaTask: print ongoing state

Commit:
AquaTask: fix started check

Commit:
Testing: fix tesing

daitenshionyan committed:
Merge pull request #2 from daitenshionyan:branch-Level-7

Branch-Level-7

Commit:
Merge remote-tracking branch 'origin/master' into branch-Level-8

Commit:
Merge branch 'branch-Level-8'

Commit:
Fix testing

Commit:
Abstract output

Commit:
Abstract user input

Commit:
Merge remote-tracking branch 'origin/add-gradle-support'

Commit:
Gradle: set main class

Commit:
Remove useless code

Commit:
Testing: add parse unit test

Commit:
TaskManager: fix toString extra new line

Commit:
Testing: add TaskManager unit tests

Commit:
Ignore IDE files

Commit:
Fix jar naming

Commit:
TaskManager: add filter function

Commit:
TaskManager: add filter test

Commit:
Command: add FIND

Commit:
Testing: add filter output tests

Commit:
ProcedureExecutor: delete

Commit:
Follow coding standards

Commit:
Add documentation

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Merge branch 'branch-Level-9'

Commit:
TaskManager: fix naming

Commit:
Relodable: add documentation

Commit:
FileUtils: add documentation

Commit:
UiManager: add documentation

Commit:
UiManager: refactor code

Commit:
DateUtils: add documentation

Commit:
TaskManager: add documentation

Commit:
Follow coding standard

Commit:
AddEventCommand: use constant variable

Commit:
build.gradle: add javafx support

Commit:
FileUtils: add getResourceUrl method

Commit:
UiComponent: create class

Commit:
Refactor code to use service and task

Commit:
Aqua: delete class

Commit:
UiManager: Refactor to be more customizable

Commit:
Executor: create class

Commit:
Manager: split logic and IO

Commit:
MainWindow: create class

Commit:
Create starter class

Commit:
MainWindow: add speech bubble

Commit:
IoManager: remove formatting

Commit:
IoManager: remove unused import

Commit:
Rename App to Aqua

Commit:
Reply: add kaomoji

Commit:
Loading: refactor loading

Commit:
Add greeting

Commit:
Command: add isLoading parameter

Commit:
Rename dispatcher to service

Commit:
Loading: refactor code

Commit:
Executor: add loading message

Commit:
MainWindow: add auto scroll

Commit:
Style: remove debugging red color

Commit:
Merge remote-tracking branch 'origin/branch-Level-10'

Commit:
Add checkstyle

Commit:
Style: use substructure

Commit:
Merge branch 'branch-A-CheckStyle'

Commit:
CheckStyle: remove trailing whitespace

Commit:
Documentation: correct documentation

Commit:
Limit method visibility

Commit:
ExecutionTask: make method name more meaningful

Commit:
Rename Executor to AppManager

Commit:
AppManager: add documentation

Commit:
CheckStyle: follow coding standards

Commit:
CheckStyle: follow coding standard

Commit:
Rename Command to ServiceProvider

Commit:
Command: create enum

Commit:
Parser: Change to use Command enum

Commit:
Merge branch 'Refactor-Enum'

Commit:
MainWindow: add documentation

Commit:
SpeechBubble: add documentation

Commit:
CheckStyle: follow coding standard

Commit:
LogicManager: add documentation

Commit:
Add documentation

Commit:
Follow coding standards

Commit:
Add citation

ChinJunAn commented on src/main/java/aqua/MainApp.java:
```diff
@@ -0,0 +1,20 @@
+public class MainApp extends Application {
+    @Override
+    public void start(Stage primaryStage) {
+        MainWindow window = new MainWindow(new LogicManager());
```
I like how you separated the definition of the GUI components into another class. This would make it easier to for future developments. Also, it enforces OOP. Good job!

ChinJunAn commented on src/main/java/aqua/aquatask/AquaDeadline.java:
```diff
@@ -0,0 +1,77 @@
+package aqua.aquatask;
+
+import java.time.LocalDateTime;
+import java.util.Optional;
```
I have noticed your usage of Optionals. I believe this is a good practice to prevent null exceptions. Also, this shows your competency in java programming

ChinJunAn commented on src/main/java/aqua/aquatask/AquaEvent.java:
```diff
@@ -0,0 +1,93 @@
+/** Implementation of AquaTask to represent an Event */
+public class AquaEvent extends AquaTask {
+    /** Tag of {@code from} argument when parsing. */
+    public static final String TAG_FROM = "from";
```
Great job following the coding standard!

ChinJunAn commented on src/main/java/aqua/logic/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,42 @@
+/**
+ * A full implementation of AddTaskCommand that creates and adds AquaDeadlines.
+ */
+public class AddDeadlineCommand extends AddTaskCommand {
```
Again, great job on practicing OOP by separating the command classes!

ChinJunAn commented on src/main/java/aqua/storage/Loadable.java:
```diff
@@ -0,0 +1,14 @@
+     *
+     * @throws LoadException if an error occured while loading.
+     */
+    public void load() throws LoadException;
```
May I check the purpose of this class?

daitenshionyan replied:
Thank you @ChinJunAn for spotting this. I think this class will only be implemented by `AppManager` throughout this whole iP.

At first I thought there might be more things to load like maybe a Config class that stores the user preference (font size, path to save directory, etc) or a LogRecord class that stores the log records and so on. So then all of these classes will implement this Loadable interface and stored into a `List` where we use a for loop or Streams to laod all of them.

```java
for (Loadable loadable : loadables) {
    loadable.load();
}
```

```java
loadables.stream().forEach(loadable -> loadable.load());
```

The `try` and `catch` block has been ommitted and `loadables` is a `List<Loadable>` that has been initialised else where (maybe the constructor).

I think the benifits of this over line by line calling the `load()` method is that we can handle the LoadException easier. For example, if we want the program to continue loading even if one class fails to load then we can do something like this.

```java
List<LoadException> loadExs = new ArrayList<>();
loadables.stream()
        .forEach(loadable -> {
                try {
                    loadable.load();
                } catch (LoadException loadEx) {
                    // just add to list and continue with next
                    loadExs.add(loadEx);
                }
            });
if (!loadExs.isEmpty()) {
    for (LoadException loadEx : loadExs) {
        loadEx.printStackTrace();
    }
} else {
    System.out.println("Loaded everything successfully");
}
```

Again, for the scale of the project I do not think these will be used.
Thank you for your comment!!

ChinJunAn reviewed (commented):
Very well-written code! I could learn from you

Commit:
Synchronize service execution

Commit:
LogicManager: make more abstract

Commit:
Merge branch 'Synchronization'

Commit:
IoManager: fix kaomoji

Commit:
SpeechBubble: add icon

Commit:
Merge branch 'Icon'

Commit:
Style: Add theme

Commit:
Add Enter to send

Commit:
Add command suggestions

Commit:
ParserService: create class

Commit:
Make parsing a queued service

If the user enters input fast enought, and some of them are invalid
inputs, the invalid input error message will race to be displayed with
the valid input result message. This may result in messages to be
displayed in the wrong order.

Let's make parsing a service that will be queued up to be executed in
the order that they are queued.

This synchronises the displyaing of command line result messages and
error messages from parsing.

Commit:
Add reply delay

Commit:
Implement history removal

Commit:
Fix enter to send

Commit:
Suggestions: fix flickering

Commit:
Change ServiceProvider to CommandController

Commit:
Command: add getSyntax method

Commit:
Suggestion: add syntax help

Commit:
CmdSyntax: add syntax

Commit:
Command: add description

Commit:
Suggestions: show description

Commit:
Style: change accent color

Commit:
Merge branch 'feature-command-suggestions'

Commit:
Documentation

Commit:
Merge branch 'feature-chat-delay'

Commit:
Remove delay

Commit:
Remove suggestion box

Commit:
MainWindow: extract method

Commit:
CommandLineParser: remove obvious comment

Commit:
UiComponent: use assertion

Commit:
Make commands display messages

Commands may have different ways of displaying messages such as a popup.
Allowing the commands to display the messages instead allows them to
display the message in the form that they desire.

Let's pull up the displaying logic from AppManager into the commands by
passing them the IoManager as well for execution.

Passing the command the AppManager instance might result in cyclic
dependencies.

Commit:
IoManager: add popup method

Commit:
IoManager: fix popup

Commit:
Command: add view schedule

Commit:
TaskManager: add filter by time

Commit:
DateSchedule: create class

Commit:
ScheduleComponent: create class

Commit:
DaySchedule: add day label

Commit:
Schedule: fix timeables spanning multiple days

Commit:
ViewCommand: add logic to show schedule

Commit:
Popup: use stage to show popup

Popup is an undecorated window. Reimplementing the decorations of this
window is difficult and error prone. Unlike a window, Popup will by
default always appear on top of all others even when unfocused.

Let's use a Stage to contain the popup display instead.

This saves many lines of code and sanity.

Commit:
Remove unused code

Commit:
Popup: set style

Commit:
Extract start of week

Commit:
Package schedule graphic classes

Commit:
Timeable: Change to abstract class

Commit:
ScheduleTimeable: create class

Commit:
Schedule: use Schedule timeable to display

There is no way to customise how a schedule for is displayed.

Let's use ScheduleTimeable instead to display schedules.

ScheduleTimeable defines methods to customise the way a schedule is
displayed. ScheduleComponent and DaySchedule are responsible for putting
the schedules on screen only and not deciding how it should be
displayed. Thus, the tasks should not be passed to those classes.
Passing the tasks will also limit the schedule graphic classes to only
work with AquaTasks which reduce resusability and abstraction.

Commit:
DaySchedule: add displaying logic

Commit:
Style: style schedule

Commit:
Schedule: move date label to ScheduleComponent

Commit:
Schedule: draw markings

Commit:
Schedule: wrap row in VBox to center

Commit:
Schedule: edit style to be round

Commit:
Schedule: add header

Commit:
Style: add row separator

Commit:
Schedule: add psudoclass style

Commit:
Schedule: add style for task completion

Commit:
Schedule: add support for deadline

Commit:
DateUtils: use isEqual to check equality

Commit:
Schedule: add style for deadline

Commit:
Schedule: separate close tasks

Commit:
Fix kaomoji not compiling

Commit:
ParserService: add documentation

Commit:
Merge branch 'fix-kaomoji'

daitenshionyan committed:
Merge pull request #6 from daitenshionyan:branch-A-CodeQuality

Branch-A-CodeQuality

daitenshionyan committed:
Merge pull request #7 from daitenshionyan:branch-A-Assertions

Branch-A-Assertions

Commit:
Merge branch 'master' of https://github.com/daitenshionyan/ip

Commit:
Merge branch 'master' into branch-B-ViewSchedules

Commit:
Schedule: fix graphics

Commit:
TaskManager: use filter report as return result

Commit:
Extract displaying logic from VIEW command

Commit:
TaskView: add todo view

Commit:
Edit GUI format

Commit:
Schedule: add tooltip to tasks

Commit:
Command VIEW: add date parameter

Commit:
Fix empty parameter value

Commit:
Schedule: add hover

Commit:
Add CI

Commit:
Merge branch 'master' into branch-B-ViewSchedules

Commit:
Add documentation

daitenshionyan committed:
Merge pull request #8 from daitenshionyan:branch-B-ViewSchedules

Branch-B-ViewSchedules

Commit:
Aqua tasks: Edit documentation

Commit:
ProcedureException: shorten name

Commit:
SyntaxException: shorten name

Commit:
Exception: edit documentation

Commit:
DaySchedule: extract block creation

Commit:
DaySchedule: simplify width calculations

Commit:
DaySchedule: extract pixel calculation

Commit:
DaySchedule: edit naming

Commit:
DaySchedule: simplify code

Commit:
Schedule: remove duplicate ROW_WIDTH

Commit:
DaySchedule: use better naming

Commit:
Chnage CI to work on master only

Commit:
UserTasks: rename classes to be more meaningful

Commit:
UserTask: fix documentation

Commit:
UserDeadline: use better naming

Commit:
UserEvent: use better naming

Commit:
UserTask: fix documentation punctuation

Commit:
UserEvent: fix documentation

Commit:
UserDeadline: fix documentation grammer

Commit:
UserTodo: fix documentation

Commit:
DaySchedule: remove class variable

Commit:
DaySchedule: add documentation

Commit:
DaySchedule: change method name

Commit:
DaySchedule: add documentation

Commit:
WeekSchedule: rename class

Commit:
WeekSchedule: add documentation

Commit:
WeekSchedule: extract to method

Commit:
WeekSchedule: add reason comments

Commit:
ScheduleHeader: add documentation

Commit:
ScheduleHeader: change visibility

Commit:
Graphics: fix documentation

Commit:
Timable: change class name

Commit:
Command: fix code quality

Commit:
Parser: fix code quality

Commit:
Logic: fix code quality

Commit:
Managers: fix code quality

Commit:
Utils: Improve code quality

Commit:
CheckStyle: fix coding standard

Commit:
AddTaskCommand: saving message not showing

Commit:
SpeechBubble: fix getRoot()

Commit:
Rename: to Something

Commit:
Rename back to original

daitenshionyan committed:
Merge pull request #9 from daitenshionyan/refactor

Refactor

Commit:
Schedule: fix clipping

Commit:
Schedule: fix hover highlighting

Commit:
Schedule: fix coding standards

Commit:
Schedule: change minimum block width

daitenshionyan committed:
Style: fix hover for schedule

daitenshionyan committed:
Schedule: remove unused pseudoclass

daitenshionyan committed:
Merge pull request #10 from daitenshionyan/fix-schedule

Fix-schedule

daitenshionyan committed:
Fix naming

daitenshionyan committed:
DateUtilsTest: add intersecting tests

daitenshionyan committed:
DateUtils: fix intersecting

daitenshionyan committed:
DateUtils: fix method naming

daitenshionyan committed:
DateUtilsTest: add toStartOfWeek test

daitenshionyan committed:
DateUtilsTest: add toStartOfDay test

daitenshionyan committed:
ArgumentParserTest: add tests

daitenshionyan committed:
ArgumentParser: edit error message

daitenshionyan committed:
Command: remove dead code

daitenshionyan committed:
AddEventCommand: handle start time after end time

daitenshionyan committed:
WeekSchedule: remove magic literals

daitenshionyan committed:
Limit visibility

daitenshionyan committed:
WriteCommand: extract method

daitenshionyan committed:
FilterCommand: extract method

daitenshionyan committed:
Add back task counter

daitenshionyan committed:
Merge branch 'branch-A-MoreTesting' into branch-release

daitenshionyan committed:
Fix checkstyle

daitenshionyan committed:
Merge 'fix-code-quality' into branch-release

daitenshionyan committed:
Fix multiple similar popups

daitenshionyan committed:
Tasks: remove onging status

daitenshionyan committed:
Add app icon

daitenshionyan committed:
Fix resizing

daitenshionyan committed:
Fix header comment

daitenshionyan committed:
Fix checkstyle

daitenshionyan committed:
Fix warning messages

daitenshionyan committed:
Split error messages

daitenshionyan committed:
Add user guide

daitenshionyan committed:
Merge pull request #12 from daitenshionyan/branch-release

Branch-release

daitenshionyan committed:
Fix fonts for mac

daitenshionyan committed:
Merge pull request #13 from daitenshionyan/fix-fonts

Fix fonts for mac
